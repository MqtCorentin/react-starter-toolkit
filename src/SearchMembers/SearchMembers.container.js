import React, { createContext } from 'react';
import PropTypes from 'prop-types';
import { LoaderModes } from 'shared/components/Loader';
import setLoaderMode from 'shared/helpers/setLoaderMode';
import SearchMembers from './SearchMembers';
import { useSearchMembers, initStateSearch } from './SearchMembers.hook';

export const SearchMembersContext = createContext({ stateSearch: initStateSearch });
const { Provider: SearchMembersProvider } = SearchMembersContext;

export const SearchMembersEnhanced = ({ useSearchMembersFn, setLoaderModeFn, ...rest }) => {
  const { anomaly, isLoading, members, submitSearch, stateSearch } = useSearchMembersFn({});

  return (
    <SearchMembersProvider value={{ stateSearch }}>
      <SearchMembers
        {...rest}
        members={members}
        loaderMode={setLoaderModeFn({ isLoading, LoaderModes })}
        anomaly={anomaly}
        submitSearch={submitSearch}
      />
    </SearchMembersProvider>
  );
};

SearchMembersEnhanced.propTypes = {
  useSearchMembersFn: PropTypes.func,
  setLoaderModeFn: PropTypes.func,
};

SearchMembersEnhanced.defaultProps = {
  useSearchMembersFn: useSearchMembers,
  setLoaderModeFn: setLoaderMode,
};

export default SearchMembersEnhanced;
