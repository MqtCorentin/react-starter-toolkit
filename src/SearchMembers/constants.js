export const TITLE_BAR = 'Recherche de membres';
export const TITLE = 'Recherche par Nom';
export const SUBTITLE = 'Tableau des membres';
export const ROUTE_URL_SEARCHMEMBERS = '/search-members';

export const FETCH_SEARCHMEMBERS = {
  INIT: 'FETCH_SEARCHMEMBERS_INIT',
  SUCCESS: 'FETCH_SEARCHMEMBERS_SUCCESS',
  FAILURE: 'FETCH_SEARCHMEMBERS_FAILURE',
};

export const TABLE_HEADERS_SEARCHMEMBERS = [
  { label: 'Pr√©nom du membre', key: 'firstname' },
  { label: 'Nom du membre', key: 'lastname' },
  { label: 'Identifiant du membre', key: '_id' },
];
