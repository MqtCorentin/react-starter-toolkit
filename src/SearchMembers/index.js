export { default, SearchMembersContext } from './SearchMembers.container';
