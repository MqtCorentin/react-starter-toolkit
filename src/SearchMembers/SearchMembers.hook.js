import { useCallback, useState, useMemo } from 'react';
import { getApi, setInitialState, useFetchData } from 'shared/helpers/fetchHook';
import setAnomalyEmptyItems from 'shared/helpers/setAnomalyEmptyItems';
import { SERVICE_NAME } from './constants';

export const INITIAL_STATE = setInitialState(SERVICE_NAME);

export const INITIAL_STATE_FORM_SEARCH_MEMBERS = {
  name: '',
  hasSubmit: false,
};

export const computeInfos = ({ members = [] }) =>
  members.map(({ _id, firstname, lastname }) => ({
    key: _id,
    firstname: {
      label: firstname,
    },
    lastname: {
      label: lastname,
    },
    _id: {
      label: _id,
    },
  }));

export const useFormSearchMembers = ({ initStateFormSearchMembers = INITIAL_STATE_FORM_SEARCH_MEMBERS }) => {
  const [stateFormSearchMembers, setStateFormSearchMembers] = useState(initStateFormSearchMembers);

  const submitFormSearchMembers = useCallback(electorsFields => {
    setStateFormSearchMembers(prevState => ({ ...prevState, ...electorsFields, hasSubmit: true }));
  }, []);

  return { submitFormSearchMembers, stateFormSearchMembers };
};

export const computeSuccess = ({ responseBody, setAnomalyEmptyItemsFn = setAnomalyEmptyItems, computeInfosFn = computeInfos }) => ({
  anomaly: {
    [SERVICE_NAME]: setAnomalyEmptyItemsFn(responseBody, 'Aucun membre ne correspond Ã  votre recherche'),
  },
  [SERVICE_NAME]: computeInfosFn({ members: responseBody }),
});

export const useSearchMembers = ({
  stateFormSearchMembers,
  initialState = INITIAL_STATE,
  serviceName = SERVICE_NAME,
  getApiFn = getApi,
  useFetchDataFn = useFetchData,
  computeSuccessFn = computeSuccess,
}) => {
  const { name = '' } = stateFormSearchMembers;
  const condition = stateFormSearchMembers.hasSubmit;

  const { state } = useFetchDataFn({
    condition,
    initialState,
    serviceName,
    computeSuccess: computeSuccessFn,
    service: useMemo(() => getApiFn(`members/search?name=${name}`), [getApiFn, name]),
  });
  return { ...state };
};
