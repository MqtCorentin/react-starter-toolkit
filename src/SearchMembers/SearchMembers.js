import React from 'react';
import Loader from 'shared/components/Loader';
import Resilience from 'shared/components/Resilience/Resilience';
import Th from 'shared/components/Table/Header/Th';
import Table from 'shared/components/Table';
import { TITLE_BAR, TITLE, SUBTITLE, TABLE_HEADERS_SEARCHMEMBERS } from './constants';
import DownloadLink from './DownloadLink';
import SearchForm from './SearchForm';

const SearchMembers = ({ header, footer, title, menu, loaderMode, members, anomaly, deleteNotification, notifications, submitSearch }) => (
  <>
    {header()}
    {menu()}
    {title({ title: TITLE_BAR })}
    <div className="af-main container">
      <h1 className="af-title--content">{TITLE}</h1>
      <SearchForm submitSearchForm={submitSearch} />
      <Loader text="Recherche des membres en cours..." mode={loaderMode}>
        <Resilience anomaly={anomaly}>
          <h2 className="af-title">{SUBTITLE}</h2>
          <Table
            items={members}
            headers={TABLE_HEADERS_SEARCHMEMBERS}
            childrenHeader={
              <Th>
                <span className="af-table__tr-label">Actions</span>
              </Th>
            }
            actionsBody={DownloadLink}
          />
        </Resilience>
      </Loader>
    </div>
    {footer()}
  </>
);

export default SearchMembers;
