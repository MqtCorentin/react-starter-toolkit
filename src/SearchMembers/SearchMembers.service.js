import { JSON_CONTENT_HEADER } from 'shared/constants';

const searchMembers = ({ fetchCustom, signal, body = {} }) =>
  fetchCustom('members', {
    signal,
    body: JSON.stringify(body),
    method: 'POST',
    headers: {
      ...JSON_CONTENT_HEADER,
    },
  });

export default searchMembers;
