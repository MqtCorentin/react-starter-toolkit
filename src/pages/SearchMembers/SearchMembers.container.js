import { func } from 'prop-types';
import setLoaderMode from 'shared/helpers/setLoaderMode';
import SearchMembers from './SearchMembers';
import { useSearchMembers, useFormSearchMembers } from './SearchMembers.hook';
import { SERVICE_NAME } from './constants';

export const SearchMembersEnhanced = ({ useSearchMembersFn, setLoaderModeFn, useFormSearchMembersFn, ...rest }) => {
  const { submitFormSearchMembers, stateFormSearchMembers } = useFormSearchMembersFn({});
  const { anomaly, isLoading, searchMembers } = useSearchMembersFn({ stateFormSearchMembers });

  return (
    <SearchMembers
      {...rest}
      members={searchMembers}
      loaderMode={setLoaderModeFn({ isLoading })}
      anomaly={anomaly[SERVICE_NAME]}
      submitSearch={submitFormSearchMembers}
    />
  );
};

SearchMembersEnhanced.propTypes = {
  useSearchMembersFn: func,
  setLoaderModeFn: func,
  useFormSearchMembersFn: func,
};

SearchMembersEnhanced.defaultProps = {
  useSearchMembersFn: useSearchMembers,
  setLoaderModeFn: setLoaderMode,
  useFormSearchMembersFn: useFormSearchMembers,
};

export default SearchMembersEnhanced;
