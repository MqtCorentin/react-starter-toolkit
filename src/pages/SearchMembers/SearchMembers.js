import { string } from 'prop-types';
import Layout from 'Layout';
import Loader from 'shared/components/Loader';
import Resilience from 'shared/components/Resilience';
import Table from 'shared/components/Table';
import DownloadLink from 'shared/components/DownloadLink';
import { formatDate } from 'shared/helpers/formatDate';
import { TITLE_BAR, TITLE, SUBTITLE, TABLE_HEADERS_SEARCHMEMBERS } from './constants';
import SearchForm from './SearchForm';

export const getDownloadPath = distributorId => `members/${distributorId}/download-detail`;

export const setFileName = ({ name = '', distributorId = '', date = new Date(), prefix = 'prefix', extension = 'csv', formatDateFn = formatDate }) =>
  `${prefix}_${name.replace(/ /g, '_').trim()}_${distributorId}_${formatDateFn(date, 'fr-CA').replace(/-/g, '')}.${extension}`;

export const DownloadLinkEnhanced = ({ idKey, firstname, lastname, getDownloadPathFn = getDownloadPath, setFileNameFn = setFileName }) => (
  <DownloadLink path={getDownloadPathFn(idKey)} fileName={setFileNameFn({ distributorId: idKey, name: `${firstname}-${lastname}` })} />
);
const SearchMembers = ({ titleBar, title, loaderMode, members, anomaly, submitSearch }) => (
  <Layout propsTitle={{ title: titleBar, backHome: true }}>
    <h1 className="af-title--content">{title}</h1>
    <SearchForm submitSearchForm={submitSearch} />
    <Loader text="Recherche des membres en cours..." mode={loaderMode}>
      <Resilience anomaly={anomaly}>
        <h2 className="af-title">{SUBTITLE}</h2>
        <Table items={members} headers={TABLE_HEADERS_SEARCHMEMBERS} />
      </Resilience>
    </Loader>
  </Layout>
);

SearchMembers.propTypes = {
  titleBar: string,
  title: string,
};

SearchMembers.defaultProps = {
  titleBar: TITLE_BAR,
  title: TITLE,
};

export default SearchMembers;
