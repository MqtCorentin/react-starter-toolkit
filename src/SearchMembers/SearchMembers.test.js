import React from 'react';
import { render } from '@testing-library/react';
import SearchMembers from './SearchMembers';

const defaultProps = {
  header: () => {},
  title: () => {},
  footer: () => {},
  menu: () => {},
  members: [],
  anomaly: null,
  loaderMode: 'none',
  submitSearch: () => {},
};

describe('<SearchMembers/>', () => {
  it('Should render SearchMembers', () => {
    const { asFragment } = render(<SearchMembers {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render <SearchMembers /> with one member', () => {
    const { asFragment } = render(
      <SearchMembers
        {...defaultProps}
        members={[
          {
            key: '1',
            name: {
              label: 'nom du membre',
            },
            id: {
              label: '1254689759',
            },
          },
        ]}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
