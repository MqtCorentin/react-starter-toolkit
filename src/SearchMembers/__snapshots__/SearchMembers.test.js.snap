// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchMembers/> Should render <SearchMembers /> with one member 1`] = `
<DocumentFragment>
  <div
    class="af-main container"
  >
    <h1
      class="af-title--content"
    >
      Recherche par Nom
    </h1>
    <form
      autocomplete="off"
      class="af-filter-inline"
      id="form-search-members"
      name="form-search-members"
    >
      <div
        class="af-filter-inline__default af-filter-inline__default--search-distributors"
      >
        <div
          class="af-filter-inline__field"
        >
          <div
            class="col-md-12"
          >
            <div
              class="af-form__text"
            >
              <label
                class="af-form__group-label"
                for="name"
              >
                Nom du membre
              </label>
              <input
                autocomplete="search-name"
                class="af-form__input-text"
                id="name"
                name="name"
                placeholder="John Doe"
                type="text"
                value=""
              />
              <small
                class="af-form__help"
              >
                Minimum 3 caractères pour un nom
              </small>
            </div>
          </div>
        </div>
        <button
          class="btn af-btn af-btn--hasiconLeft af-btn--submitconfirm af-btn--disabled"
          disabled=""
          form="form-search-members"
          type="submit"
        >
          <span
            class="af-btn__text"
          >
            Rechercher
          </span>
          <i
            class="glyphicon glyphicon-search"
          />
        </button>
      </div>
    </form>
    <div
      class="af-loader"
    >
      <h2
        class="af-title"
      >
        Tableau des membres
      </h2>
      <table
        class="af-table"
      >
        <thead
          aria-label="table-header"
          class="af-table__thead"
        >
          <tr
            aria-label="table-header-line"
            class="af-table__tr"
          >
            <th
              aria-label="table-body-line"
              class="af-table__th"
              role="columnheader"
            >
              <span
                class="af-table__tr-label"
              >
                Prénom du membre
              </span>
            </th>
            <th
              aria-label="table-body-line"
              class="af-table__th"
              role="columnheader"
            >
              <span
                class="af-table__tr-label"
              >
                Nom du membre
              </span>
            </th>
            <th
              aria-label="table-body-line"
              class="af-table__th"
              role="columnheader"
            >
              <span
                class="af-table__tr-label"
              >
                Identifiant du membre
              </span>
            </th>
            <th
              aria-label="table-body-line"
              class="af-table__th"
              role="columnheader"
            >
              <span
                class="af-table__tr-label"
              >
                Actions
              </span>
            </th>
          </tr>
        </thead>
        <tbody
          aria-label="table-body"
          class="af-table__body"
        >
          <tr
            aria-label="table-body-line"
            class="af-table__tr"
          >
            <td
              class="af-table__cell"
            >
              nom du membre
            </td>
            <td
              class="af-table__cell"
            >
              1254689759
            </td>
            <td
              class="af-table__cell af-table__cell--actions"
            >
              <div
                class="af-loader af-loader--fullscreen"
              >
                <button
                  aria-label="download-link"
                  class="af-link af-link--hasIconLeft af-link--download"
                  type="button"
                >
                  <svg
                    alt="download-csv"
                    aria-labelledby="title"
                    class="af-download-link__icon"
                    role="img"
                    viewBox="0 0 100 100"
                  >
                    <title>
                      download-csv
                    </title>
                    <path
                      d="M70.4 5H13.9v90h72.2V20.7L70.4 5zM83 91.2H50v-7.8h24.7V67.8H49.9v-5.2L63 39.8h-6.8V16.9H43.5v22.9h-6.8l13.1 22.7v5.2H25v15.6h24.8v7.8H16.7V8.4h53.8v12.3H83v70.5zM49.7 76.3l-1-.4c-.8-.3-1.6-1-1.6-2.2 0-1.3 1.2-2.4 2.9-2.4.9 0 1.9.3 2.6 1l-1 1.3c-.5-.4-1-.6-1.6-.6-.5 0-.8.2-.8.5 0 .4.4.5 1.2.8l.9.4c1 .4 1.5 1.1 1.5 2.1 0 1.3-1.1 2.5-3 2.5-1 0-2.1-.4-3-1.1l1.2-1.4c.6.4 1.2.8 1.8.8.6 0 .9-.2.9-.6 0-.3-.4-.4-1-.7zm3.4-4.8h2.2l.8 3.4c.2.8.4 1.6.6 2.5.2-.8.4-1.6.6-2.5l.8-3.4h2.1l-2.3 7.8h-2.5l-2.3-7.8zm-9 6.1c.6 0 1.1-.3 1.4-.7l1.1 1.3c-.7.8-1.6 1.2-2.6 1.2-2 0-3.7-1.3-3.7-4 0-2.6 1.8-4.1 3.8-4.1 1 0 1.9.5 2.4 1l-1.1 1.3c-.4-.3-.8-.6-1.3-.6-.9 0-1.7.8-1.7 2.3 0 1.5.7 2.3 1.7 2.3z"
                    />
                  </svg>
                  <span
                    class="af-link__text"
                  >
                    Télécharger
                  </span>
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SearchMembers/> Should render SearchMembers 1`] = `
<DocumentFragment>
  <div
    class="af-main container"
  >
    <h1
      class="af-title--content"
    >
      Recherche par Nom
    </h1>
    <form
      autocomplete="off"
      class="af-filter-inline"
      id="form-search-members"
      name="form-search-members"
    >
      <div
        class="af-filter-inline__default af-filter-inline__default--search-distributors"
      >
        <div
          class="af-filter-inline__field"
        >
          <div
            class="col-md-12"
          >
            <div
              class="af-form__text"
            >
              <label
                class="af-form__group-label"
                for="name"
              >
                Nom du membre
              </label>
              <input
                autocomplete="search-name"
                class="af-form__input-text"
                id="name"
                name="name"
                placeholder="John Doe"
                type="text"
                value=""
              />
              <small
                class="af-form__help"
              >
                Minimum 3 caractères pour un nom
              </small>
            </div>
          </div>
        </div>
        <button
          class="btn af-btn af-btn--hasiconLeft af-btn--submitconfirm af-btn--disabled"
          disabled=""
          form="form-search-members"
          type="submit"
        >
          <span
            class="af-btn__text"
          >
            Rechercher
          </span>
          <i
            class="glyphicon glyphicon-search"
          />
        </button>
      </div>
    </form>
    <div
      class="af-loader"
    >
      <h2
        class="af-title"
      >
        Tableau des membres
      </h2>
    </div>
  </div>
</DocumentFragment>
`;
